#SNP Filtering Exercise for B@G 2016 Winter School
Designed by Jon Puritz

#GOALS
1. To learn how to use VCFtools to filter a VCF file for missing data, genotype depth, locus quality score, minor allele frequency, and genotype call depth
2. To learn how to use vcflib to filter FreeBayes VCF files generated with RAD data
3. To filter a VCF file for HWE within populations
4. How to decompose a VCF into SNPs and INDELs and
5. How to use a haplotyping script to further filter SNPs for paralogs and genotyping errors.
***
###Start of Exercise
Welcome to the SNP filtering exercise.  For the first part of the exercise, the filtering steps should work on almost any VCF file.  
For the second part of the exercise, we are going to assume you are working with a VCF file that was generated by
FreeBayes.  Note that other SNP callers can be configured to include the same annotations.
Let's find our way back to your original working directory and make a new filtering directory
```bash
cd ~/D1W
mkdir filtering
cd filtering
module load dDocent/v2.18
```
Now, let's download some data to look at.  
Building haplotypes for WL_080
Building haplotypes for WL_032
Building haplotypes for WL_071
Building haplotypes for WL_081
Building haplotypes for BR_004
Building haplotypes for BR_021
Building haplotypes for BR_015
Building haplotypes for BR_043
Building haplotypes for WL_066
Filtered 26 loci below missing data cutoff
Filtered 66 possible paralogs
Filtered 17 loci with low coverage or genotyping errors
Filtered 0 loci with an excess of haplotypes
```
The script found another 109 loci to remove from our file.  Besides this output to the terminal, the script outputs a file called stats.out

```bash
head stats.out
Locus		Sites	Haplotypes	Inds_Haplotyped	Total_Inds	Prop_Haplotyped	Status	Poss_Paralog	Low_Cov/Geno_Err	Miss_Geno	Comment
E10001_L101	1		2			30				31			0.968			PASSED	0				0					1	
E10003_L101	7		9			30				31			0.968			PASSED	1				0					0	
E10004_L101	-		-			-				-			-				FILTERED0				0					1			Complex
E10008_L101	1		2			30				31			0.968			PASSED	0				0					1	
E10013_L142	3		6			30				31			0.968			PASSED	0				0					1	
E10014_L117	2		3			31				31			1.000			PASSED	0				0					0	
E10024_L101	1		2			31				31			1.000			PASSED	0				0					0	
E10029_L101	1		2			31				31			1.000			PASSED	0				0					0	
```
We can use this file to create a list of loci to filter

```bash
grep FILTERED stats.out | mawk '!/Complex/' | cut -f1 > loci.to.remove
```
Now that we have the list we can parse through the VCF file and remove the bad RAD loci
I've made a simple script to do this remove.bad.hap.loci.sh

```bash
curl -L -O https://github.com/jpuritz/dDocent/raw/master/scripts/remove.bad.hap.loci.sh
chmod +x remove.bad.hap.loci.sh
./remove.bad.hap.loci.sh loci.to.remove SNP.DP3g95p5maf05.HWE.recode.vcf 
```
This produces a FINAL FINAL FINAL filtered VCF file	SNP.DP3g95p5maf05.HWE.filtered.vcf

```bash
mawk '!/#/' SNP.DP3g95p5maf05.HWE.filtered.vcf | wc -l
```
We're left with 7,666 SNPs!
How many possible errors?

```bash
./ErrorCount.sh SNP.DP3g95p5maf05.HWE.filtered.vcf
```

```bash
This script counts the number of potential genotyping errors due to low read depth
It report a low range, based on a 50% binomial probability of observing the second allele in a heterozygote and a high range based on a 25% probability.
Potential genotyping errors from genotypes from only 1 read range from 0 to 0.0
Potential genotyping errors from genotypes from only 2 reads range from 0 to 0.0
Potential genotyping errors from genotypes from only 3 reads range from 302 to 1014.72
Potential genotyping errors from genotypes from only 4 reads range from 162 to 822.232
Potential genotyping errors from genotypes from only 5 reads range from 88 to 669
31 number of individuals and 7666 equals 237646 total genotypes
Total genotypes not counting missing data 237081
Total potential error rate is between 0.00232831816974 and 0.0105700245908
SCORCHED EARTH SCENARIO
WHAT IF ALL LOW DEPTH HOMOZYGOTE GENOTYPES ARE ERRORS?????
The total SCORCHED EARTH error rate is 0.0330857386294.

```

**Congrats!  You've finished the Filtering Exercise**

#Even bigger Congrats!  You've finished all the Exercises for Day 1!

###One extra note: My lab is currently writing up a manuscript on the rad_haplotyper script, so please contact me before sharing it with anyone.  Thanks!
